internal/controllers/extension_controller.go

```diff
@@ -0,0 +1,177 @@
+
+	ocv1alpha1 "github.com/operator-framework/operator-controller/api/v1alpha1"
+	"github.com/operator-framework/operator-controller/internal/controllers/validators"
+)
+
+// ExtensionReconciler reconciles a Extension object
+type ExtensionReconciler struct {
+	client.Client
+	BundleProvider BundleProvider
+	Scheme         *runtime.Scheme
+	Resolver       *solver.Solver
+}
+
+//+kubebuilder:rbac:groups=olm.operatorframework.io,resources=extensions,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=olm.operatorframework.io,resources=extensions/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=olm.operatorframework.io,resources=extensions/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *ExtensionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx).WithName("extension-controller")
+	l.V(1).Info("starting")
+	defer l.V(1).Info("ending")
+
+	var existingExt = &ocv1alpha1.Extension{}
+	if err := r.Client.Get(ctx, req.NamespacedName, existingExt); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	reconciledExt := existingExt.DeepCopy()
+	res, reconcileErr := r.reconcile(ctx, reconciledExt)
+
+	// Do checks before any Update()s, as Update() may modify the resource structure!
+	updateStatus := !equality.Semantic.DeepEqual(existingExt.Status, reconciledExt.Status)
+	updateFinalizers := !equality.Semantic.DeepEqual(existingExt.Finalizers, reconciledExt.Finalizers)
+	unexpectedFieldsChanged := r.checkForUnexpectedFieldChange(*existingExt, *reconciledExt)
+
+	if updateStatus {
+		if updateErr := r.Client.Status().Update(ctx, reconciledExt); updateErr != nil {
+			return res, utilerrors.NewAggregate([]error{reconcileErr, updateErr})
+		}
+	}
+
+	if unexpectedFieldsChanged {
+		panic("spec or metadata changed by reconciler")
+	}
+
+	if updateFinalizers {
+		if updateErr := r.Client.Update(ctx, reconciledExt); updateErr != nil {
+			return res, utilerrors.NewAggregate([]error{reconcileErr, updateErr})
+		}
+	}
+
+	return res, reconcileErr
+}
+
+// Compare resources - ignoring status & metadata.finalizers
+func (*ExtensionReconciler) checkForUnexpectedFieldChange(a, b ocv1alpha1.Extension) bool {
+	a.Status, b.Status = ocv1alpha1.ExtensionStatus{}, ocv1alpha1.ExtensionStatus{}
+	a.Finalizers, b.Finalizers = []string{}, []string{}
+	return !equality.Semantic.DeepEqual(a, b)
+}
+
+// Helper function to do the actual reconcile
+//
+// Today we always return ctrl.Result{} and an error.
+// But in the future we might update this function
+// to return different results (e.g. requeue).
+//
+//nolint:unparam
+func (r *ExtensionReconciler) reconcile(_ context.Context, ext *ocv1alpha1.Extension) (ctrl.Result, error) {
+	// Don't do anything if Paused
+	if ext.Spec.Managed == ocv1alpha1.ManagedStatePaused {
```

Comment:
Can we add a log here stating no reconciliation done because the managedState is set to paused?

Reply:
For follow-up, check in other popular Kubernetes projects to and see if there is a pattern where a Paused condition is set.